src/solvers/dg_triangles/analysis.jl

```diff
@@ -0,0 +1,76 @@
+function calc_error_norms(func, u::StructArray, t, analyzer,
+                          mesh::AbstractMeshData{Dim}, equations, initial_condition,
+                          dg::DG{<:RefElemData{Dim}}, cache, cache_analysis) where {Dim}
+  rd = dg.basis
+  md = mesh.md
+  @unpack u_values = cache
+
+  # interpolate u to quadrature points
+  StructArrays.foreachfield(mul_by!(rd.Vq), u_values, u) 
+
+  # convert md.xyz::NTuple{Dim,Matrix} to StructArray for broadcasting
+  xyzq = StructArray{SVector{Dim,real(dg)}}(md.xyzq)
+  u_exact_values = initial_condition.(xyzq, t, equations) # todo: move to cache
+
+  # `pointwise_error` is a StructArray{SVector{nvariables(equations),real(dg)}}, so to square each entry 
+  # we need to apply a double broadcast (x->x.^2). 
+  pointwise_error = func.(u_values, equations) - func.(u_exact_values, equations) # todo: for loop to avoid allocations
+  component_l2_errors = sum(md.wJq .* (x->x.^2).(pointwise_error)) 
+  component_linf_errors = maximum((x->abs.(x)).(pointwise_error))
+
+  return component_l2_errors, component_linf_errors
+end
+
+function integrate(func::Func, u,
+                   mesh::AbstractMeshData,
+                   equations, dg::DG{<:RefElemData}, cache; normalize=true) where {Func}
+  rd = dg.basis
+  md = mesh.md
+  @unpack u_values = cache
+
+  # interpolate u to quadrature points
+  StructArrays.foreachfield(mul_by!(rd.Vq), u_values, u) 
+
+  integral = sum(md.wJq .* func.(u_values, equations))
+  if normalize == true
+    integral /= sum(md.wJq)
+  end
+  return integral
+end
+
+function analyze(::typeof(entropy_timederivative), du, u, t,
+                 mesh::AbstractMeshData, equations, dg::DG{<:RefElemData}, cache)
+
+  rd = dg.basis
+  md = mesh.md
+  @unpack u_values = cache
+
+  # interpolate u, du to quadrature points
+  du_values = similar(u_values) # todo: move to cache
+  StructArrays.foreachfield(mul_by!(rd.Vq), du_values, du) 
+  StructArrays.foreachfield(mul_by!(rd.Vq), u_values, u) 
+
+  # compute ∫v(u) * du/dt = ∫dS/dt. We can directly compute v(u) instead of 
+  # computing the entropy projection here, since the RHS will be projected 
+  # to polynomials of degree N and testing with the L2 projection of v(u) 
+  # would be equivalent to testing with v(u) due to the moment-preserving 
+  # property of the L2 projection.
+  dS_dt = zero(real(dg))
+  for i in Base.OneTo(length(md.wJq)) 
+    dS_dt += dot(cons2entropy(u_values[i],equations), du_values[i]) * md.wJq[i]
+  end
+  return dS_dt
+end
+
+function create_cache_analysis(analyzer, mesh::AbstractMeshData,
+                               equations, dg::DG{<:RefElemData}, cache,
+                               RealT, uEltype)
+  return (; )               
+end
+
+## todo: what else is required in the interface for AnalysisCallback?
+SolutionAnalyzer(rd::RefElemData) = rd
+
+# can we add mesh as an argument to this?
+nelements(solver::DG{<:RefElemData}, cache) = size(cache.u_values,2) 
```

Comment:
Sure

Reply:
I think similar changes might need to be made in the `analysis_callback` code. I haven't checked closely, will leave this conversation open until I can make those.

Reply:
https://github.com/trixi-framework/Trixi.jl/blob/bdacf612e817881e12f1a4a80da465df57ab8f82/src/callbacks_step/analysis.jl#L191 

I'm hesitant to change this line b/c it seems like it might break things for other solvers.

Reply:
I guess that's just a lesson that we learned from your PR: Methods like `nelements` etc. should also get the `mesh` as additional argument everywhere. Thus, we would use `nelements(mesh, solver, cache)`, `eachelement(mesh, solver, cache)` etc.

Reply:
Should we make that a separate PR?

Reply:
I'm fine with doing it here to verify that it's also working for your new solvers.